adapters/oidc/js/src/main/resources/keycloak.d.ts

```diff
@@ -292,13 +292,23 @@ declare namespace Keycloak {
 	}
 
 	interface KeycloakTokenParsed {
+		acr?: string;
+		aud?: string;
+		auth_time?: number;
+		azp?: string;
+		email_varified?: boolean;
 		exp?: number;
 		iat?: number;
+		iss?: string;
+		jti?: string;
 		nonce?: string;
-		sub?: string;
-		session_state?: string;
+		preferred_username?: string;
 		realm_access?: KeycloakRoles;
 		resource_access?: KeycloakResourceAccess;
+		scope?: string;
+		session_state?: string;
+		sub?: string;
+		typ?: string;

```

Comment:
As a token can contain custom claims would it not be better to just define this as a map?

Reply:
Yeah, ideally this would be a thing the user can provide using a generic. For example:

```ts
interface MyTokenFields {
  foo: string;
}

const tokenParsed = keycloak.getParsedToken<MyTokenFields>();

tokenParsed.foo; // TypeScipt knows this property throught the generic
```

But since we have none of these methods exist, and these properties are exposed publicly this not a feasible solution. So instead we can specify only the properties here that are guaranteed to be present at all time (@stianst any idea what those might be?).

Next to that we can specify an [index signature](https://www.typescriptlang.org/docs/handbook/2/objects.html#index-signatures) so that TypeScript users can themselves determine what is inside of the token. So the interface would look a bit like this:

```ts
export interface KeycloakTokenParsed {
  alwaysPresent: string;
  [key: string]: any;
}
```

Now if someone wants to get a value from the token it should be fine for TypeScript:

```ts
keycloak.tokenParsed.foo // This property will be of type 'any' 
```

Now if someone wants to narrow this down they can use type predicate.

```ts
if (typeof keycloak.tokenParsed.foo === 'string') {
  keycloak.tokenParsed.foo // This property will be of type 'string'
}
```

Reply:
The last example you gave looks like the way to go for now. That's what I had in mind when I said "map". There's some built-in claims, but probably always will be custom ones as well.